tests/log-test.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const assert = require('assert');
+
+const MockRequire = require('mock-require');
+
+const AWSMock = require('./../mocks/aws-s3-mock');
+
+MockRequire('aws-sdk', AWSMock);
+
+process.env.AWS_ACCESS_KEY_ID = 'S3RVER';
+
+process.env.AWS_SECRET_ACCESS_KEY = 'S3RVER';
+
+const Log = require('./../index');
+
+const LogError = require('./../lib/log-error');
+
+const fakeLog = {
+	type: 1,
+	entity: 'api',
+	entity_id: 'product',
+	message: '[GET] Request from 0.0.0.0 a product/custom_data',
+	date_created: 1559103066,
+	user_created: 0,
+	log: {
+		uri: {
+			controller: 'product',
+			action: 'custom_data',
+			data: []
+		},
+		responseHttpCode: 200,
+		responseTime: '0.3236'
+	},
+	id: 'a1d2asd1-1a23-a23d-as1d-0asdas2130'
+};
+
+describe('Log', () => {
+
+	beforeEach(() => {
+		AWSMock.S3.restore();
+	});
+
+	it('should not reject when put a log into S3', async () => {
+		assert.doesNotThrow(() => { Log.add(fakeLog, 'someBucket'); });
+	});
+
+	it('should not reject and generate the log id when put a log into S3 without id', async () => {
+
+		const fakeLogWithoutId = { ...fakeLog };
+		delete fakeLogWithoutId.id;
+		assert.doesNotThrow(() => { Log.add(fakeLogWithoutId, 'someBucket'); });
+
+	});
+
+	it('should throw LogError when S3 throws', async () => {
+
+		AWSMock.S3.throws = true;
+		assert.throws(() => { Log.add({ id: 1 }, 'someBucket'); }, {
+			name: 'LogError',
+			code: LogError.codes.S3_ERROR
+		});
+
+	});
+
+	it('should throw LogError when the Log is invalid or is empty', async () => {
```

Comment:
Acá se esta testeando que throwee cuando sea un array vacío, el description no es correcto.
Debe ser algo como "... when the log given is an empty array"

Tema 2, faltan muchos casos de error:
- `Log.add(true, 'fooBucket');`
- `Log.add(54, 'fooBucket');`
- `Log.add(['foo', 'bar'], 'fooBucket');`
- `Log.add(null, 'fooBucket');`

-

Reply:
Para no tener que hacer multiples casos bloques `it()`, podes meter los casos invalidos en un array, recorrerlo y probar por cada valor.
En la description, podrias agrupar como "... when Log is not an object"